Text/Megaparsec/Error.hs

```diff
@@ -248,6 +252,98 @@ parseErrorPretty
 parseErrorPretty e =
   sourcePosStackPretty (errorPos e) <> ":\n" <> parseErrorTextPretty e
 
+
+-- | Pretty-print a 'ParseError' and display the line on which the parse error
+-- occurred. The rendered 'String' always ends with a newline.
+--
+-- Assumes that the stream tokens can be coerced to an @Int@ and that the value
+-- integer value @10@ corresponds to a new-line character.
+--
+-- @since 6.0.0
+
+parseErrorPrettyWithLine
+  :: ( Enum (Token s)
+     , Ord t
+     , ShowToken t
+     , ShowErrorComponent e
+     , Stream s )
+  => s              -- ^ Original input stream
+  -> ParseError t e -- ^ Parse error to render
+  -> String         -- ^ Result of rendering
+parseErrorPrettyWithLine stream e = mconcat messageComponents
+  where
+    position = errorPos e
+    lineMessage = renderLineMessage stream $ NE.last position
+    messageComponents =
+      [ sourcePosStackPretty position
+      , ":\n"
+      , lineMessage 
+      , parseErrorTextPretty e ] 
+
+-- | Renders the source context to improve error message quality.
+
+renderLineMessage
+  :: forall s .
+     ( Enum (Token s)
+     , Stream s )
+  => s -> SourcePos -> String
+renderLineMessage s p = unlines [ paddingLine, contextLine, pointingLine ]
+  where
+    paddingLine   = padding <> "|"
+    contextLine   = " " <> lineNumberStr <> " | " <> cs
+    pointingLine  = padding <> "|" <> replicate columnNum ' ' <> "^"
+
+    lineNumberStr = show $ unPos linePos
+    padLength = length lineNumberStr + 2
+    padding = replicate padLength ' '
+
+    cs =
+      case ts of
```

Comment:
Two things I'm not sure about:

1. Is printing `"<empty line>"` better than just printing empty line?
2. How to handle the cases when we point to a `'\n'` character? Probably we should place the caret `^` to the end of line the `'\n'` character ends? This is a corner case we should test.